ignite/metrics/multilabel_confusion_matrix.py

```diff
@@ -0,0 +1,112 @@
+import numbers
+from typing import Callable, Optional, Sequence, Tuple, Union
+
+import torch
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+from ignite.metrics.metrics_lambda import MetricsLambda
+
+__all__ = [
+    "MultiLabelConfusionMatrix",
+]
+
+
+class MultiLabelConfusionMatrix(Metric):
+    def __init__(
+        self,
+        num_classes: int,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+        normalized: bool = False,
+    ):
+        if num_classes <= 1:
+            raise ValueError("Argument num_classes needs to be > 1")
+
+        self.num_classes = num_classes
+        self._num_examples = 0
+        self.normalized = normalized
+        super(MultiLabelConfusionMatrix, self).__init__(output_transform=output_transform, device=device)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self.confusion_matrix = torch.zeros(self.num_classes, 2, 2, dtype=torch.int64, device=self._device)
+        self._num_examples = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        self._check_input(output)
+        y_pred, y = output[0].detach(), output[1].detach()
+
+        self._num_examples += y.shape[0]
+        y_reshaped = y.swapaxes(0, 1).reshape(self.num_classes, -1)
+        y_pred_reshaped = y_pred.swapaxes(0, 1).reshape(self.num_classes, -1)
+        y_total = torch.unsqueeze(torch.count_nonzero(y_reshaped, axis=1), 0)
+        y_pred_total = torch.unsqueeze(torch.count_nonzero(y_pred_reshaped, axis=1), 0)
+        tp = torch.unsqueeze(torch.count_nonzero(torch.mul(y_reshaped, y_pred_reshaped), axis=1), 0)
+        fp = y_pred_total - tp
+        fn = y_total - tp
+        tn = y_reshaped.shape[1] - tp - fp - fn
+
+        self.confusion_matrix += torch.cat([tn, fp, fn, tp], dim=0).T.reshape(-1, 2, 2)
```

Comment:
Can we directly create the tensor in the good shape and device (`self._device`) ? For the shape I'm not yet sure, but device will be necessary.

Reply:
You mean running the concatenation operation on self._device?

Reply:
There are two points:
- can we do `torch.cat([tn, fp, fn, tp], dim=0).T.reshape(-1, 2, 2)` in less ops ?
- I think there is a device mismatch if `self.confusion_matrix` is on CPU and tp, tn, ... are on GPU.

Reply:
First point : We should not be able to do it any faster. I think we are good here since this should not be slower than: 
```
tp = torch.unsqueeze(torch.count_nonzero(torch.mul(y_reshaped, y_pred_reshaped), axis=1), 0)
fp = y_pred_total - tp
fn = y_total - tp
tn = y_reshaped.shape[1] - tp - fp - fn
```
Second point : In case the original ```y``` and ```y_pred``` are not on the same device as ```self._device```, should we use ```.to(self._device)``` on both of them? That should ensure no device mismatch.

Reply:
Second point: the idea of `self._device` is to have the option where to store internal things: CPU or GPU. If user provides y_preds and y on X device, let's perform all the compurations on this X device and only the final op will convert the result to the required `self._device` (to avoid device mismatch).

First point: sounds good. We can come back to this point later if there are any other ideas on how to improve the performances.